### ISSUE TITLE
MustAsync makes browser hanging on without resolving

### ISSUE THREAD

**Issue by mdmoura** [JeremySkinner/FluentValidation#81]
Hello,

I have the following Validator on a MVC5 site:

```
public class CatalogCreateModelValidator : AbstractValidator<CatalogCreateModel> {

  public CatalogCreateModelValidator(Func<IMediator> mediator) {      
    RuleFor(x => x.Name)
      .NotEmpty().WithMessage("The name is required")
      .MustAsync(name => mediator().SendAsync(new CatalogNameValidationModel(name)))
      .WithMessage("The name is unavailable");
  }
}
```

When the name is empty I get an error on my view.
When name is not empty the browser hangs on "loading" without showing the error.

It seems it hangs on MustAsync ... The validation of the CatalogNameValidationModel is:

```
public class CatalogNameValidationModelHandler : IAsyncOrderHandler<CatalogNameValidationModel, Boolean> {

  private readonly IDbContext _context;

  public CatalogNameValidationModelHandler(IDbContext context) {
    _context = context;
  }

public async Task<Boolean> Handle(CatalogNameValidationModel model) {
   return !await _context.Catalogs.AnyAsync(x => x.Name == model.Name);
} // Handle
```

  }

My application works fine with Async in controllers ...

The only think it is not working is the MustAsync validation.

Thank You,
Miguel

**Comment by JeremySkinner** [comment]
Could you put together a unit test that reproduces the problem? I'll then be able to look at it further.

**Comment by mdmoura** [comment]
I just created a simple project based on ASP.NET MVC template project where I was able to replicate the error: https://github.com/mdmoura/FluentValidation.Async

The validator is the following:

  public class HomeIndexModelValidator : AbstractValidator<HomeIndexModel> {
    public HomeIndexModelValidator() {
      RuleFor(x => x.Name)
        .NotEmpty().WithMessage("The name is required")
        .MustAsync(x => new Task<Boolean>(() => { Task.Delay(1000); return true; })).WithMessage("Must Async validation is false");
    }
  }

If you start the project you will get a form.
If you submitted it without filling with will get the message "The name is required".
If you fill the name then the browser hangs out when, i think, must async is being called.

Any idea why this happens?

**Comment by JeremySkinner** [comment]
Thanks - I'll take a look at this and try to get back to you in the next couple of weeks.

**Comment by mdmoura** [comment]
That would be great! I have two applications using Async which break with FV's MustAsync.

Thank You

**Comment by JeremySkinner** [comment]
So the problem is that you're trying to invoke an async validator in a synchronous way.

Using the built-in MVC validation always runs FluentValidation synchronously (essentially calling validator.Validate), but you need to invoke it asynchronously by manually getting a reference to the validator and then calling ValidateAsync. 

That being said, there *is* a bug - if you call the validator synchronously, it shouldn't hang like this - it should fall back to running MustAsync validator as though it were a standard Must call. But in your case this isn't what you want - you'll still need to call ValidateAsync if you want async behaviour.

Hope that makes sense

Jeremy

**Comment by mdmoura** [comment]
Yes, it makes a lot of sense. 

With async controller MVC should include something like ModelState.IsValidAsync. No?

I am going to take a look at ASP.NET MVC VNext and try to figure if this is the same.

I am considering creating two Action Filters which gets the Validator through Injection.
ValidateModelFilter would call Validate and ValidateModelAsync would call ValidateAsync.

This way I would not use ModelState.IsValid. I would only add the errors to ModelState in the filters.

**Comment by JeremySkinner** [comment]
ModelState.IsValid doesn't actually perform validation - it just checks to see if errors have already been generated, so a ModelState.IsValidAsync wouldn't make sense. 

Validation actually occurs during model binding, so an UpdateModelAsync or something like that would be needed. 

I don't know if filters support async code either, so you'd need to look into that.

**Comment by mdmoura** [comment]
> Validation actually occurs during model binding, so an UpdateModelAsync or something like that would be needed. 

Yes, you are right. I am looking at Asp.Net VNext to check if anything changed.

When do you expect to have FluentValidation for Mvc6?

> That being said, there is a bug - if you call the validator synchronously, it shouldn't hang like this - it should fall back to running MustAsync validator as though it were a standard Must call.

When do you expect to fix this bug? It would be useful because it would allow the same validator to work as Async and Sync without the need of having two validators when there is that need.

**Comment by JeremySkinner** [comment]
> When do you expect to have FluentValidation for Mvc6?

It's a work in progress, but won't be finished until MVC6 is released. The Validation API for MVC6 isn't actually finished yet - so I'm waiting for that to stabalise before I do any more.

> When do you expect to fix this bug? It would be useful because it would allow the same validator to work as Async and Sync without the need of having two validators when there is that need.

No time soon I'm afraid - I'm very busy with my job at the moment, which has to take priority over my OSS work. You're welcome to submit a pull request though - I'm much more likely to get a pull-request merged in quickly.

**Comment by sshumakov** [comment]
Hello Jeremy,

I've noticed deadlock on WPF app when I call ValidateAsync which has MustAsync method inside it.

Not sure why this happen, on server side everything is okay.

Are you aware of any issues with MustAsync in combination with ValidateAsync?

**Comment by JeremySkinner** [comment]
I'm not aware of any issues, but I personally have very little experience with the async code in FV.

**Comment by sshumakov** [comment]
It turned out that I had another non async call to Validate method. This caused UI thread hanging.

**Comment by damiangreen** [comment]
Is there a way I can annotate my validator or configure the factory so that MVC knows to invoke it async?

Currently I can't have a validator;
    public class LoginModelValidator : AbstractValidator<Domain.Models.LoginModel>
    {
        private readonly IUserService _userService;

        public LoginModelValidator(IUserService userService)
        {
            _userService = userService;

            RuleFor(x => x.UserName)
                .Cascade(CascadeMode.StopOnFirstFailure)
                .NotEmpty()
                .MustAsync(async name => await _userService.UserNameExists(name))
..etc

and then in my Ioc Registration

    FluentValidationModelValidatorProvider.Configure(
                provider =>
                {
                    provider.ValidatorFactory = new ValidatorFactory();
                    provider.AddImplicitRequiredValidator = false;
                });

I can't seem to see any injection point to force async validation. It just freezes.

**Comment by JeremySkinner** [comment]
No, sorry, that's not something that's currently supported. You'd need to invoke the validator manually by running ValidateAsync rather than relying on the automatic integration.

**Comment by damiangreen** [comment]
Thanks for your reply Jeremy, from your experience is this a restriction of the MVC framework, i.e. no injection points, or just something that needs building into your (excellent) framework?

**Comment by JeremySkinner** [comment]
I think it could be extended to support this - probably using IValidatorInterceptor - I'd need to take some time to investigate this, but probably not for a while I'm afraid.

**Comment by kamranayub** [comment]
Damn I just got bitten by this. In my case the async method was throwing an exception and it just hung the browser. It wasn't clear/intuitive that I need to call a different validate method if I add an async rule, I just assumed it was taken care of. Regardless, it shouldn't hang.

**Comment by mdmoura** [comment]
@kamranayub I think this is not solved yet ...

**Comment by kamranayub** [comment]
@mdmoura it's not, but in my case it's okay since I don't use the integrated MVC validation, I run custom validators against commands.

**Comment by mdmoura** [comment]
@kamranayub Out of curiosity, are you using Mediatr with Messages and FluentValidation interceptor?

**Comment by kamranayub** [comment]
@mdmoura No, I haven't heard of Mediatr. This is a custom command pattern we're using.

**Comment by siulung4073** [comment]
+1

**Comment by petekcchen** [comment]
Encountered the same hanging issue and it happens only when I need async/await syntax inside MustAsync

for instance, the following wont hang the browser
.MustAsync((request, username, cancellationToken) => userRepository.UsernameExistsAsync(username))

but the following will hang the browser
.MustAsync(async (request, username, cancellationToken) => !await userRepository.UsernameExistsAsync(username))

Any idea how I can resolve this? Thanks.

**Comment by sshumakov** [comment]
petekcchen, try to see if you have non async calls to Validate method. This is usually the reason for that behavior. Also, try to add .ConfigureAwait(false) to userRepository.UsernameExistsAsync method.

**Comment by maca88** [comment]
I've created a PR that can reslove the deadlock problem by using a custom async predicate validator (example included in the PR).

**Comment by damiangreen** [comment]
I've been looking at the commit history and I can't tell if this made it in. Is this in a beta release yet?

**Comment by maca88** [comment]
[A fix](https://github.com/JeremySkinner/FluentValidation/blob/master/src/FluentValidation/Validators/AsyncValidatorBase.cs#L24) for this issue was applied and is also available in the latest beta release.

**Comment by damiangreen** [comment]
Thanks @maca88 So i can use it as I showed in the example above , i.e now now deadlock?  I dont need to create any kind of async validator or anything?

**Comment by maca88** [comment]
You can achieve this by saving the tasks for getting your min and max values in the constructor and use them when validating. I created a [simple example](https://dotnetfiddle.net/dOKEsp). As you can see in the example, the min and max value are retrieved only once.

**Comment by damiangreen** [comment]
@maca88 Brilliant. Thanks for that

**Comment by velkumars** [comment]
Hi, 
Could some one please help me to resolved this? i'm trying to change **CustomAsync to MustAsync**, but i couldn't make things to work. Below is my custom method 

**Trying to use MustAsync**

  ```
RuleFor(o => o).MustAsync(o => {
              return  CheckIdNumberAlreadyExist(o)
            });

        private static async Task<ValidationFailure> CheckIdNumberAlreadyExist(SaveProxyCommand command)
        {
            if (command.Id > 0)
                return null;

            using (IDbConnection connection = new SqlConnection(ConnectionSettings.LicensingConnectionString))
            {
                var param = new DynamicParameters();
                param.Add("@idnumber", command.IdNumber);

                var vehicle = await connection.QueryFirstOrDefaultAsync<dynamic>("new_checkDuplicateProxyIdNumber", param, commandType: CommandType.StoredProcedure);

                return vehicle != null
                    ? new ValidationFailure("IdNumber", "Id Number Already Exist")
                    : null;
            }
         }
```

**Comment by JeremySkinner** [comment]
@velkumars This is an old thread. Please open a separate issue for your problem, and also provide full details of what's not working including any error messages. Thanks.
